rgw: Introduce BucketEncryption APIs to support SSE-S3 feature

rahuldevp opened the pull request:
This patch introduces support for 3 new BucketEncryption APIs which are listed
below and are helpful in supporting AWS SSE-S3 encryption mode.
PutBucketEncryption: https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketEncryption.html
GetBucketEncryption: https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketEncryption.html
DeleteBucketEncryption: https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteBucketEncryption.html

The user provided parameters are parsed and stored in the bucket's extended
attributes RGW_ATTR_BUCKET_ENCRYPTION and
RGW_ATTR_BUCKET_ENCRYPTION_SSE_S3_KEY_ID.

Signed-off-by: Rahul Dev Parashar rahul.dev@flipkart.com


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

cbodley commented on src/rgw/rgw_common.h:
```diff
@@ -143,6 +144,10 @@ using ceph::crypto::MD5;
 #define RGW_ATTR_CRYPT_DATAKEY  RGW_ATTR_CRYPT_PREFIX "datakey"
 
+/* SSE-S3 Encryption Attributes */
+#define RGW_ATTR_BUCKET_ENCRYPTION RGW_ATTR_PREFIX "encryption."
+#define RGW_ATTR_BUCKET_ENCRYPTION_SSE_S3_KEY_ID RGW_ATTR_BUCKET_ENCRYPTION "sse-s3-key-id"
```
the trailing `.` seems odd for the `"encryption."` policy attribute

this looks similar to `#define RGW_ATTR_CRYPT_PREFIX   RGW_ATTR_PREFIX "crypt."` above, but we don't use `RGW_ATTR_CRYPT_PREFIX` directly as an attribute name

how about something like this?
```c++
#define RGW_ATTR_BUCKET_ENCRYPTION_PREFIX RGW_ATTR_PREFIX "sse-s3."
#define RGW_ATTR_BUCKET_ENCRYPTION_POLICY RGW_ATTR_BUCKET_ENCRYPTION_PREFIX "policy"
#define RGW_ATTR_BUCKET_ENCRYPTION_KEY_ID RGW_ATTR_BUCKET_ENCRYPTION_PREFIX "key-id"
```

rahuldevp replied:
This looks better Casey. I will make the necessary changes along with other suggestions.

cbodley commented on src/rgw/rgw_common.h:
```diff
@@ -39,6 +39,7 @@
 #include "cls/rgw/cls_rgw_types.h"
 #include "include/rados/librados.hpp"
 #include "rgw_public_access.h"
+#include "rgw_bucket_encryption.h"
```
it doesn't look like `rgw_common.h` depends on anything from this header

mattbenjamin commented:
this is great progress, @rahuldevp

rahuldevp commented:
> this is great progress, @rahuldevp

Thanks @mattbenjamin

rahuldevp commented on src/rgw/rgw_op.cc:
```diff
@@ -8510,3 +8510,126 @@ void RGWDeleteBucketPublicAccessBlock::execute(optional_yield y)
+
+void RGWGetBucketEncryption::execute(optional_yield y)
+{
+  auto attrs = s->bucket_attrs;
```
@cbodley Need a little more clarification about this change Casey. Do you mean that I should replace 
`auto attrs = s->bucket_attrs;
  if (auto aiter = attrs.find(RGW_ATTR_BUCKET_ENCRYPTION);
      aiter == attrs.end()) {`
with
`auto aiter = s->bucket_attrs.find(RGW_ATTR_BUCKET_ENCRYPTION_POLICY);
  if (aiter == s->bucket_attrs.end()) {`
or any other suggestion is also there?

rahuldevp replied:
@cbodley Can you please confirm on this if my above comment is the right way or you were suggesting something else?

cbodley replied:
the only part that needs to change is this:
```diff
-  auto attrs = s->bucket_attrs;
+  const auto& attrs = s->bucket_attrs;
```

rahuldevp commented on src/rgw/rgw_kms.cc:
```diff
@@ -159,16 +159,35 @@ add_name_val_to_obj(const char *n, std::string &v, rapidjson::GenericValue<E,A>
+  virtual std::string & auth() = 0;
+  virtual std::string & k_namespace() = 0;
+  virtual std::string & prefix() = 0;
+  virtual const std::string & secret_engine() = 0;
```
This change was pushed for implementing SSE-S3 but later reverted to be raised as a separate PR. So resolving this conversation as it is no longer relevant here.

rahuldevp commented on src/rgw/rgw_bucket_encryption.cc:
```diff
@@ -0,0 +1,41 @@
+void ApplyServerSideEncryptionByDefault::decode_xml(XMLObj *obj) {
+  RGWXMLDecoder::decode_xml("KMSMasterKeyID", kmsMasterKeyID, obj, false);
+  if(kmsMasterKeyID.compare("") != 0) {
+    throw RGWXMLDecoder::err("implementation for KMS is not supported yet");
```
Moved checks to `RGWPutBucketEncryption::execute()` and specific error codes are returned. For "aws:kms" and algorithm other than AES256 we are returning `ERR_NOT_IMPLEMENTED` with the respective message.

rahuldevp commented on src/rgw/rgw_bucket_encryption.h:
```diff
@@ -0,0 +1,130 @@
+
+  string sse_algorithm() const {
+    return sseAlgorithm;
+  }
```
Done

rahuldevp commented:
@cbodley I have addressed review comments and fixed checks which were failing. I had to squash and merge old commits as an old commit was failing Signed-off-by check.

cbodley commented:
thanks @rahuldevp, could you please add some simple test cases of these APIs to https://github.com/ceph/s3-tests/?

prasakri commented:
> thanks @rahuldevp, could you please add some simple test cases of these APIs to https://github.com/ceph/s3-tests/?

https://github.com/ceph/s3-tests/pull/404

cbodley commented:
👍 https://pulpito.ceph.com/cbodley-2021-07-26_21:18:53-rgw-wip-cbodley-testing-distro-basic-smithi/

> 2021-07-26T22:46:01.694 INFO:teuthology.orchestra.run.smithi101.stderr:s3tests_boto3.functional.test_s3.test_put_bucket_encryption ... ok
2021-07-26T22:46:02.619 INFO:teuthology.orchestra.run.smithi101.stderr:s3tests_boto3.functional.test_s3.test_get_bucket_encryption ... ok
2021-07-26T22:46:03.552 INFO:teuthology.orchestra.run.smithi101.stderr:s3tests_boto3.functional.test_s3.test_delete_bucket_encryption ... ok
